### PR TITLE
Fix for crash when displaying collection of strings

### PR DESCRIPTION
**Pull request by itowlson**
If a task depends on a set of task IDs, then Batch Explorer crashes when attempting to display that task.  The immediate cause is that when recursing into the collection, BE tries to expand the properties of type String, which include the indexed property Chars, so GetValue crashes with "invalid parameter count".  But even excluding indexed properties just leaves it displaying Length instead of the string content.  This fix has BE display "stringable" collection elements as strings, just like normal properties, instead of always attempting to recurse into collection elements.

### CODE REVIEW

**Comment by itowlson** [comment]
Aargh, it has that blasted upstream/master merge again!  Grr!  I believe you can squash that as part of merging the PR, but if not, get back to me and I'll recreate it.

**Comment by paselem** [comment]
+Matthew


Change looks good to me. Unless Matt has any other concerns I'm happy accepting the pull.


-Pablo


Sent from Outlook<http://aka.ms/weboutlook>

________________________________

**Comment by matthchr** [comment]
@itowlson @paselem I will look at this today and merge it if I don't have any comments

**Comment by matthchr** [comment]
@itowlson Rather than duplicating the code in the foreach for handling enumerables, should we instead refactor the recursion to be the following?
1) Check if `o` is null and if so return null (i.e. skip the object).
2) Check if `o` has a `ToString` defined and if so, use that.
3) If no `ToString` is defined, do the `TimeSpan` check we have now, and the `Enumerable` check we have now.
4) If none of the above, recurse on each property.

i.e. make the recursion `object` focused and not property focused?

**Comment by matthchr** [comment]
Something like this:
```
        protected List<PropertyModel> ObjectToPropertyModel(object o, List<string> propertiesToOmit = null)
        {
            if (propertiesToOmit == null)
            {
                propertiesToOmit = new List<string>();
            }
            propertiesToOmit.Add("CustomBehaviors");

            List<PropertyModel> result = ExtractPropertyCollection(o, propertiesToOmit);
            result.Add(new SimplePropertyModel(LastUpdateFromServerString, this.LastUpdatedTime.ToString()));
            return result;
        }

        private static List<PropertyModel> ExtractPropertyCollection(object o, List<string> propertiesToOmit = null)
        {
            List<PropertyModel> results = new List<PropertyModel>();
            PropertyInfo[] properties = o.GetType().GetProperties();

            Func<PropertyInfo, PropertyModel> extractPropertyModelFunc = prop =>
                {
                    try
                    {
                        object propertyValue = prop.GetValue(o);
                        return ObjectToPropertyModelRecursive(prop.Name, propertyValue, propertiesToOmit);
                    }
                    catch (TargetInvocationException)
                    {
                        // Skip. Cannot access all properties for bound objects
                    }

                    return null;
                };

            results = properties
                .Where(prop => propertiesToOmit == null || !propertiesToOmit.Contains(prop.Name))
                .Select(extractPropertyModelFunc)
                .Where(model => model != null).ToList();

            return results;
        }

        private static PropertyModel ObjectToPropertyModelRecursive(string propertyName, object o, List<string> propertiesToOmit = null)
        {
            Type objectType = o == null ? typeof(object) : o.GetType();
            Type toStringDeclaringType = objectType.GetMethod("ToString", Type.EmptyTypes).DeclaringType;
            PropertyModel propertyModel = null;

            if (o == null)
            {
                //Don't track null properties in order to keep the list short
                //propertyModel = new SimplePropertyModel(propInfo.Name, string.Empty);
            }
            else if ((objectType == typeof(TimeSpan) || objectType == typeof(TimeSpan?)) && o.Equals(TimeSpan.MaxValue))
            {
                propertyModel = new SimplePropertyModel(propertyName, "Unlimited");
            }
            //For other types which have a ToString declared, we want to use the built in ToString()
            else if (toStringDeclaringType != typeof(object))
            {
                propertyModel = new SimplePropertyModel(propertyName, o.ToString());
            }
            //If we have a collection, enumerate the contents
            else if (typeof(IEnumerable).IsAssignableFrom(objectType))
            {
                IEnumerable enumerable = o as IEnumerable;
                string prefix = CollectionPropertyModel.GetItemDisplayPrefix(enumerable);
                List<PropertyModel> collectionModel = new List<PropertyModel>();
                int i = 0;
                foreach (object item in enumerable)
                {
                    PropertyModel innerPropertyModel = ObjectToPropertyModelRecursive(prefix + i, item, propertiesToOmit);
                    collectionModel.Add(innerPropertyModel);
                    i++;
                }

                propertyModel = new CollectionPropertyModel(propertyName, collectionModel);
            }
            //For any complex properties without a ToString of their own, we recurse
            else
            {
                propertyModel = new CollectionPropertyModel(propertyName, ExtractPropertyCollection(o, propertiesToOmit));
            }

            return propertyModel;
        }
```

**Comment by itowlson** [comment]
Great suggestions @matthchr. I've incorporated them with a few changes.

File: CSharp/BatchExplorer/Helpers/EnumerableExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Microsoft.Azure.BatchExplorer.Helpers
+{
+    internal static class EnumerableExtensions
+    {
+        internal static IEnumerable<T> SelectMany<T>(this IEnumerable source, Func<object, int, IEnumerable<T>> selector)
```
Comment (matthchr) [inline CSharp/BatchExplorer/Helpers/EnumerableExtensions.cs]: Shouldn't this `object` be a `T`?

Comment (matthchr) [inline CSharp/BatchExplorer/Helpers/EnumerableExtensions.cs]: Doesn't SelectMany have `TSource` and `TResult` usually?

File: CSharp/BatchExplorer/Models/ModelBase.cs [hunk 2]
```diff
@@ -71,85 +72,78 @@ protected List<PropertyModel> ObjectToPropertyModel(object o, List<string> prope
...
+            {
+                object propertyValue = property.GetValue(o);
+                return ObjectToPropertyModelRecursive(property.Name, propertyValue, propertiesToOmit);
+            }
+            catch (TargetInvocationException)
+            {
+                // Skip. Cannot access all properties for bound objects
+                return Enumerable.Empty<PropertyModel>();
+            }
+
```
Comment (matthchr) [inline CSharp/BatchExplorer/Models/ModelBase.cs]: minor: extra space

File: CSharp/BatchExplorer/Helpers/EnumerableExtensions.cs (hunk 1 above)
Comment (itowlson) [inline CSharp/BatchExplorer/Helpers/EnumerableExtensions.cs]: If I had an IEnumerable<TSource> then I wouldn't need to write a custom extension.

**Pull request merged**